### PR TITLE
Build pip package

### PR DESCRIPTION
**Pull request by PeterVenable**

### CODE REVIEW

**Comment by PeterVenable** [comment]
@ejball is there a good way to put some secret configuration into the build process? Publishing the pip package requires a secret token.

**Comment by ejball** [comment]
@PeterVenable You can set [secret environment variables](https://github.com/FacilityApi/FacilityPython/settings/secrets).

File: BuildPip.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Building the FacilityPython Pip Package
+
+When `facility.py` changes, a new verison of [`facilitypython`](https://pypi.org/project/facilitypython/) should be published.
+
+1. Update the version number in `setup.py` to match `VersionPrefix` in `Directory.build.props`.
+2. Ensure PyPi credentials with permission to update `facilitypython` are in `~/.pypirc` or ready for interactive entry.
```
Comment (ejball) [inline BuildPip.md]:
```suggestion
2. Ensure PyPI credentials with permission to update `facilitypython` are in `~/.pypirc` or ready for interactive entry.
```

File: README.md [hunk 2]
```diff
@@ -11,3 +11,17 @@ Facility.CodeGen.Python | A library that generates Python for a Facility Service
...
+which is included here and also published
+as a Python package on PyPI at
+https://pypi.org/project/facilitypython/
+
+To install it:
+```
+pip install facilitypython
+```
+
+Also see: [Building the facilitypython pypi package](BuildPip.Md)
```
Comment (ejball) [inline README.md]:
```suggestion
Also see: [Building the facilitypython pypi package](BuildPip.md)
```

File: pip/setup.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from setuptools import setup, find_packages
+
+setup(
+    name='facilitypython',
+    version='0.11.0',
+    author="Peter Venable",
+    author_email="peter.venable@gmail.com",
+    description="Supports FacilityAPI web service clients generated by fsdgenpython.",
```
Comment (ejball) [inline pip/setup.py]:
```suggestion
    description="Supports Facility API Framework web service clients generated by fsdgenpython.",
```

File: pip/setup.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    long_description_content_type="text/plain",
+    long_description="Contains facility.py which is imported by web service clients generated by fsdgenpython from FacilityAPI specifications.",
```
Comment (ejball) [inline pip/setup.py]:
```suggestion
    long_description="Contains facility.py which is imported by web service clients generated by fsdgenpython from Facility Service Definitions.",
```

**Comment by PeterVenable** [comment]
Where is the build defined? I'd like to update it to support publishing the pip package. Currently, running `build pip-publish` will attempt to publish it, but fail unless the version has been bumped since the last time it was published. How does the Nuget publish process deal with that?

File: tools/Build/Build.cs [hunk 5]
```diff
@@ -50,5 +68,38 @@ void CodeGen(bool verify)
 
 			RunDotNet(toolPath, "conformance/ConformanceApi.fsd", "conformance/", "--newline", "lf", verifyOption);
 		}
+
+		void CreatePipPackage()
+		{
+			CopyFiles("src", "pip/facilitypython", "*.py");
+			AppRunner.RunApp("python", "-m pip install --user --upgrade setuptools wheel twine".Split());
```
Comment (ejball) [inline tools/Build/Build.cs]:
```suggestion
			RunApp("python", "-m pip install --user --upgrade setuptools wheel twine".Split());
```

Add `using static Faithlife.Build.DotNetRunner;` up top.

**Review by ejball — APPROVED** [review]
Can we use GitHub Actions to publish automatically, like the NuGet packages?

**Comment by ejball** [comment]
@PeterVenable The build is defined here: https://github.com/FacilityApi/FacilityPython/blob/master/.github/workflows/build.yaml

**Comment by ejball** [comment]
> How does the NuGet publish process deal with that?

It has a `--skip-duplicates` option. There are probably better ways to handle it, but I've just ignored all npm errors for the same effect, e.g. `npm publish || true`.

File: tools/Build/Build.cs (hunk 5 above)
Comment (PeterVenable) [reply tools/Build/Build.cs]: Maybe that's the wrong `using` statement. I get `The name 'RunApp' does not exist in the current context`

Comment (ejball) [reply tools/Build/Build.cs]: Sorry, I meant `using static Faithlife.Build.AppRunner;`

Comment (PeterVenable) [reply tools/Build/Build.cs]: Ah, it's `using static Faithlife.Build.AppRunner;`

**Pull request merged**